service/mapreduce_service: Cancel query when stopping

dawmd opened the pull request:
Before these changes, shutting down a node could be prolonged because of mapreduce_service. `mapreduce_service::stop()` uninitializes messaging service, which includes waiting for all ongoing RPC handlers. We already had a mechanism for cancelling local mapreduce tasks, but we were missing one for cancelling external queries.

In this commit, we modify the signature of the request so it supports cancelling via an abort source. We also provide a reproducer test for the problem.

Fixes scylladb/scylladb#22337

Backport: not needed. This is an enhancement.

avikivity commented on idl/mapreduce_request.idl.hh:
```diff
@@ -45,5 +45,5 @@ struct mapreduce_result {
 };
 
-verb mapreduce_request(query::mapreduce_request req [[ref]], std::optional<tracing::trace_info> trace_info [[ref]]) -> query::mapreduce_result;
+verb [[cancellable]] mapreduce_request(query::mapreduce_request req [[ref]], std::optional<tracing::trace_info> trace_info [[ref]]) -> query::mapreduce_result;
```
@gleb-cloudius does this change the wire encoding in any way?

gleb-cloudius replied:
No. This only cancels local sender or waiter. It does not cancel the operation on a server and thus does not change the protocol.

avikivity replied:
@dawmd how does it work then? cancellation only works if the RPC wasn't sent yet.

avikivity replied:
@dawmd what about this point? How does cancellation work if the query started?

avikivity replied:
Does it help? Won't most RPCs be in progress and therefore won't be canceled?

avikivity replied:
I see. What about remote work that someone else sent us?

dawmd replied:
Those tasks will also be cancelled. It's already implemented:

https://github.com/scylladb/scylladb/blob/3bb19e9ac98a8aab42620724771e2b7d67562875/service/mapreduce_service.cc#L480-L493

As soon as the node receives a signal to shut down, we set `_shutdown` to true. The supercoordinator that sent us the tasks will receive `rpc::closed_error` as the response and will attempt to perform the query locally.

avikivity replied:
ok

avikivity commented on service/mapreduce_service.cc:
```diff
@@ -364,6 +367,11 @@ future<query::mapreduce_result> mapreduce_service::dispatch_to_shards(
 ) {
+    while (utils::get_local_injector().enter("prevent_local_mapreduce_tasks")) {
+        flogger.warn("Preventing dispatching a mapreduce task to the shards due to error injection 'prevent_local_mapreduce_tasks'. Sleeping for 5 seconds");
+        co_await seastar::sleep(std::chrono::seconds(5));
```
This makes the test 5 seconds longer, no?

dawmd replied:
Replied here: https://github.com/scylladb/scylladb/pull/22651#discussion_r1939296289
I guess the reactor cancels the sleep immediately once we get the signal.
I chose 5 seconds arbitrarily to not overly flood the logs. Maybe changing it to like

```cpp
if (utils::get_local_injector().enter("prevent_local_mapreduce_tasks") {
  // logging
  while (true) { co_await seastar::sleep(std::chrono::milliseconds(100)); }
}
```

would be better.

piodul commented on test/topology_custom/test_aggregation.py:
```diff
@@ -0,0 +1,61 @@
+        # We use this timeout because it's longer than the cumulative timeout of the following
+        # steps. For the test to be reliable, the query cannot end on its own and that's also
+        # why we're intentionally discarding the future here.
+        _ = cql.run_async("SELECT count(*) FROM ks.t BYPASS CACHE USING TIMEOUT 600s", host=host1)
```
AFAIK runaway tasks are bound to cause problems in our test framework, we have some code that detects whether there are some tasks running at the end of the test and mark it as failed if there are some. Moreover, I'm not sure whether the query will run at all if you don't await it - IIRC Python coroutines are lazy and do nothing unless you poll them. You should make sure that you poll it and wait until it completes (and ignore the error).

I recommend taking a look into asyncio's [TaskGroups](https://docs.python.org/3/library/asyncio-task.html#task-groups), they would be a good fit here.

avikivity commented on service/mapreduce_service.hh:
```diff
@@ -141,7 +142,10 @@ public:
-        , _early_abort_subscription(as.subscribe([this] () noexcept { _shutdown = true; }))
+        , _early_abort_subscription(as.subscribe([this] () noexcept {
+            _shutdown = true;
+            _as.request_abort();
```
Please give better names. We have `as` triggering `_as`, but I can't tell what either one relates to.

Using short names stops working when you have multiple objects of the same class, the names then have to reflect the role of the object.

avikivity commented on test/topology_custom/test_aggregation.py:
```diff
@@ -0,0 +1,61 @@
+        await s2_log.wait_for("Preventing dispatching a mapreduce task to the shards due to error injection " \
+                              "'prevent_local_mapreduce_tasks'. Sleeping for 5 seconds", from_mark=s2_mark, timeout=60)
+        # Verify that the supercoordinator stops without an issue despite the ongoing mapreduce task.
+        await manager.server_stop_gracefully(s1.server_id, timeout=120)
```
How long does the test take to run?

dawmd replied:
10 seconds. Decreasing the sleep in the error injection loop from 5 to 1 seconds doesn't change that value.

avikivity commented on test/topology_custom/test_aggregation.py:
```diff
@@ -0,0 +1,61 @@
+    outgoing queries to other nodes, which would otherwise prevent the shutdown.
+    """
+
+    [s1, s2] = await manager.servers_add(2, cmdline=["--logger-log-level", "forward_service=trace"])
```
Can't you reuse a cluster from a pool rather than create a new one? Like cqlpy tests reuse the cluster.

dawmd replied:
I haven't seen a cqlpy test stop a node and I assumed the suite wasn't designed for this kind of steps. If that's possible and okay, though, I'll move the test there.

dawmd commented:
Version 3:
* Moved the test to the topology suite to reuse the cluster. One of the notes must still be initialized separately to enable additional logging, but that should decrease the time required for setting up the nodes at least a little bit.
* Changed the error injection to an infinite loop. The message is only printed once and the sleep has been reduced from 5 seconds to 100 milliseconds.

Jadw1 commented on test/topology/test_aggregation.py:
```diff
@@ -0,0 +1,74 @@
+            # We use this timeout because it's longer than the cumulative timeout of the following
+            # steps. For the test to be reliable, the query cannot end on its own.
+            try:
+                await cql.run_async("SELECT count(*) FROM ks.t BYPASS CACHE USING TIMEOUT 600s", host=host1)
```
If we expect the query to fail, shouldn't we fail the test when the query success?

dawmd replied:
Good point

Jadw1 commented on test/topology/test_aggregation.py:
```diff
@@ -0,0 +1,74 @@
+logger = logging.getLogger(__name__)
+
+@pytest.mark.asyncio
+async def test_cancel_mapreduce(manager: ManagerClient):
```
You need to add `@skip_mode` attribute, the test uses error injection which is disabled in release mode.

dawmd commented on service/mapreduce_service.cc:
```diff
@@ -364,6 +367,11 @@ future<query::mapreduce_result> mapreduce_service::dispatch_to_shards(
 ) {
+    if (utils::get_local_injector().enter("prevent_local_mapreduce_tasks")) {
+        flogger.warn("Preventing dispatching a mapreduce task to the shards due to error injection 'prevent_local_mapreduce_tasks'");
+        while (true) { co_await seastar::sleep(std::chrono::milliseconds(100)); }
```
Yes. That's a good idea, all the more because this form of injection logs a message with level INFO. Nice, thanks Michał!

dawmd commented:
Version 4:
* The ugly if/loop with error injection replaced by a brief `inject` call with a message.
* `@skip_mode("release", ...)` added.
* We fail the test if the query succeeds.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14936/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14936/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: topology/test_aggregation
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/14286/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/3153/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/6273/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14936/)
#### Build Details:
- Duration: 5 hr 13 min
- Builder: spider2.cloudius-systems.com

piodul commented on test/topology/test_aggregation.py:
```diff
@@ -0,0 +1,80 @@
+    running_servers = await manager.running_servers()
+    assert len(running_servers) > 0
+
+    s1 = await manager.server_add(cmdline=["--logger-log-level", "forward_service=trace"])
+    s2 = running_servers[0]
+    assert s1.server_id != s2.server_id
```
I'm pretty sure there is a way to change the log level in runtime. The `topology` suite already provides more than one node, so instead of adding a new node you could just assert that there are at least two nodes and change the log level of one of them.

IIRC it's the `/storage_service/logging_level` endpoint (see `api/api-doc/storage_service.json`)

piodul commented on service/mapreduce_service.cc:
```diff
@@ -364,6 +367,8 @@ future<query::mapreduce_result> mapreduce_service::dispatch_to_shards(
     query::mapreduce_request req,
     std::optional<tracing::trace_info> tr_info
 ) {
+    co_await utils::get_local_injector().inject("mapreduce_pause_dispatch_to_shards", utils::wait_for_message(10min));
```
I'm not sure what is the timeout set for our tests - it's probably shorter than 10 minutes - but what will happen if we don't provide the message before `utils::wait_for_message`'s timeout? Will the node crash, or will it just continue?

If it crashes then we can probably reduce the timeout here so that the test doesn't hang for so long if it fails. I would put one or two minutes here instead.

dawmd replied:
Good point. I went with 5 minutes to avoid some weird cases when all of the calls in the test almost use their timeout (which adds up to 4 minutes).

avikivity replied:
It's okay to be generous with timeouts that expire when the test fails, since that happens rarely. The effort should go into reducing flakiness.

dawmd commented:
Version 5:
* Implemented the functionality changing the log level of a logger at runtime. Provided a test verifying the changes.
* Replaced the manually created keyspace and table with calls to `new_test_{keyspace,table}`.
* Reduced the timeout of the error injection from 10 to 5 minutes.
* The aggregation test doesn't create any new nodes anymore.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14955/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14955/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: rest_api/test_storage_service
:small_blue_diamond: topology/test_aggregation
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/3165/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/6285/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/14299/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14955/)
#### Build Details:
- Duration: 5 hr 30 min
- Builder: spider2.cloudius-systems.com

dawmd commented:
> #### 🔴 CI State: FAILURE
> 
> ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14955/) ✅ - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14955/) **The following new/updated tests ran 100 times for each mode:** 🔹 rest_api/test_storage_service 🔹 topology/test_aggregation ✅ - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/3165/) ✅ - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/6285/) ✅ - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/14299/) ❌ - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14955/)
> #### Build Details:
> 
>     * Duration: 5 hr 30 min
> 
>     * Builder: spider2.cloudius-systems.com

scylladb/scylladb#22628

dawmd commented:
Version 6:
* Resolved the merge conflict.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15004/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15004/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: rest_api/test_storage_service
:small_blue_diamond: topology/test_aggregation
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/6323/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/3203/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/14338/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15004/)
#### Build Details:
- Duration: 5 hr 46 min
- Builder: spider8.cloudius-systems.com

dawmd commented:
> #### 🔴 CI State: FAILURE
> 
> ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15004/) ✅ - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15004/) **The following new/updated tests ran 100 times for each mode:** 🔹 rest_api/test_storage_service 🔹 topology/test_aggregation ✅ - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/6323/) ✅ - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/3203/) ✅ - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/14338/) ❌ - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15004/)
> #### Build Details:
> 
>     * Duration: 5 hr 46 min
> 
>     * Builder: spider8.cloudius-systems.com

scylladb/scylladb#22628 again

avikivity commented on service/mapreduce_service.cc:
```diff
@@ -282,9 +283,11 @@ class retrying_dispatcher {
 
         // Try to send this mapreduce_request to another node.
         try {
+            flogger.trace("Sending mapreduce request to {}", id);
```
In general, if you add trace logs, try to make them informative, for example by including `req` in the output.

They're also candidates for tracing, which is more useful since it can be selectively enabled on a request.

We can ignore it for now.

dawmd replied:
Since the CI failed, I fixed that right away. I wanted to keep those logs short because before calling the function we already log more detailed information. I added them mostly to make sure that they're right before sending a request, but I guess that's a bit too obsessive

avikivity replied:
It's fine to add logs during development and traditional to forget to remove them afterwards.

dawmd commented:
> #### 🔴 CI State: FAILURE
> 
> ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15040/) ✅ - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15040/) **The following new/updated tests ran 100 times for each mode:** 🔹 topology/test_aggregation ✅ - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/3242/) ✅ - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/6359/) ✅ - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/14374/) ❌ - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15040/)
> #### Failed Tests (1/38134):
> 
>     * [test_topology_ops[True].debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15040/testReport/junit/topology_custom/test_topology_ops/Tests___Unit_Tests___test_topology_ops_True__debug_1) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_topology_ops%5BTrue%5D.debug.1)
> 
> 
> #### Build Details:
> 
>     * Duration: 9 hr 56 min
> 
>     * Builder: i-09649faed2a7dec67 (m5ad.8xlarge)

scylladb/scylladb#22551

dawmd commented:
Version 8:
* Got rid of the redundant logs.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15131/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15131/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: topology/test_aggregation
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/14447/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/3313/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/6430/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15131/)
#### Failed Tests (1/38191):
- [test_topology_ops[False].debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15131/testReport/junit/topology_custom/test_topology_ops/Tests___Unit_Tests___test_topology_ops_False__debug_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_topology_ops[False].debug.3)
#### Build Details:
- Duration: 6 hr 20 min
- Builder: i-0f2a315b27d83bd24 (m5d.12xlarge)

dawmd commented:
> #### 🔴 CI State: FAILURE
> 
> ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15131/) ✅ - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15131/) **The following new/updated tests ran 100 times for each mode:** 🔹 topology/test_aggregation ✅ - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/14447/) ✅ - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/3313/) ✅ - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/6430/) ❌ - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15131/)
> #### Failed Tests (1/38191):
> 
>     * [test_topology_ops[False].debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15131/testReport/junit/topology_custom/test_topology_ops/Tests___Unit_Tests___test_topology_ops_False__debug_3) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_topology_ops%5BFalse%5D.debug.3)
> 
> 
> #### Build Details:
> 
>     * Duration: 6 hr 20 min
> 
>     * Builder: i-0f2a315b27d83bd24 (m5d.12xlarge)

scylladb/scylladb#22551 again. Can we merge this? I'll send a fix to prevent any more of these failures

avikivity commented:
Queued, promising a fix works
